src/expr/src/scalar.rs

```diff
@@ -1573,6 +1577,29 @@ impl MirScalarExpr {
                             **list_create = MirScalarExpr::literal_null(typ);
                         }
                     }
+                    // Peel one layer off of each of the earlier element types.
+                    for t in earlier_list_create_types.iter_mut() {
+                        if let ScalarType::List {
+                            element_type,
+                            custom_id: _,
+                        } = t
+                        {
+                            **t = *element_type.clone();
+                            // These are not the same types anymore, so remove custom_ids all the
```

Comment:
I don't follow this logic (but I also don't know exactly how `custom_id`s affect how we understand/use a type.

In the abstract, `[custom(a), custom(b), custom(c)][2]` should have `list[custom]` as its type, and after reducing we should have `custom(b)` with `custom` as the type... no?

Reply:
You are right, we might need more complicated logic here. But I'd like to fix this in a follow-up PR. The current PR definitely makes the types less wrong than they are on `main`, and it seems an exotic special case when there are custom types involved.

(Also, it might be not incorrect to even just erase _all custom ids everywhere in MIR_, similarly to how we want to erase differences among various string types in the representation types work. This is because the actual Datums that we produce in `ListCreate` are not affected by the custom ids, as far as I can tell)

Reply:
Oh my. 🫣

This is very reasonable, but let's make sure we actually sort out the story. Perhaps another argument for a separate notion of "representation type" https://github.com/MaterializeInc/materialize/pull/27239.